### PR TITLE
Parses json before attempting to do things

### PR DESCRIPTION
**Pull request by jtslear**
- addresses #4

### CODE REVIEW

File: web/controllers/webhook_controller.ex [hunk 1]
```diff
@@ -2,12 +2,17 @@ defmodule Bigben.WebhookController do
...
+        Bigben.BuildStatusUpdater.update(params["payload"])
+      _ ->
+        text conn, "Failing Builds are not needed here"
+    end
   end
 
-  def create(conn, params) do
-    text conn, "Failing Builds are not needed here"
+  defp get_status(params) do
+    payload = params["payload"] |> Poison.decode!
```
Comment (drapergeek) [inline web/controllers/webhook_controller.ex]: The payload comes in as a map already, it seems that doing this here would be overkill unless I'm missing something.

Comment (jtslear) [inline web/controllers/webhook_controller.ex]: I disagree.  You are being fed url encoded json as an input from the webhook.

File: test/controllers/webhook_controller_test.exs [hunk 2]
```diff
@@ -7,14 +7,9 @@ defmodule Bigben.WebhookControllerTest do
   alias Bigben.Build
 
   test "POST /webhook with a passing build, creates an entry", %{conn: conn} do
+  payload = ~s/{"id":112129326,"repository":{"id":1294851,"name":"findit"},"number":"84","status":0,"result":0,"status_message":"Passed","result_message":"Passed","started_at":"2016-02-23T01:20:00Z","finished_at":"2016-02-23T01:21:00Z","duration":60,"branch":"master"}/
```
Comment (drapergeek) [inline test/controllers/webhook_controller_test.exs]: Two things:
1. How about breaking this by line so we can see it well?
2. What do you think about removing the non-important stuff like `name` `number`, etc? Basically remove all things that we don't test for later.

File: web/controllers/webhook_controller.ex [hunk 3]
```diff
@@ -2,12 +2,14 @@ defmodule Bigben.WebhookController do
...
-
   def create(conn, params) do
-    text conn, "Failing Builds are not needed here"
+    json_params = params["payload"] |> Poison.decode!
+    case json_params["status"] do
+      0 ->
+        Bigben.BuildStatusUpdater.update(json_params)
+        text conn, "Success"
+      _ ->
+        text conn, "Failing Builds are not needed here"
```
Comment (drapergeek) [inline web/controllers/webhook_controller.ex:12]: :+1: 

Is there a test that covers this case?

File: web/models/build_status_updater.ex [hunk 4]
```diff
@@ -19,8 +19,8 @@ defmodule Bigben.BuildStatusUpdater do
     Map.put(params, "total_time", total_time)
   end
 
-  def get_or_create_branch(%{"branch" => branch}) do
-    Repo.get_by(Branch, name: branch) || create_branch(branch)
+  def get_or_create_branch(json_params) do
+    Repo.get_by(Branch, name: json_params["branch"]) || create_branch(json_params["branch"])
```
Comment (drapergeek) [inline web/models/build_status_updater.ex]: Why did this need to change?

File: web/controllers/webhook_controller.ex (hunk 3 above)
Comment (jtslear) [inline web/controllers/webhook_controller.ex:12]: Indeed:
https://github.com/jtslear/bigben/blob/06d13914c64fb5d411568e6920ae3e2f3e560ca9/test/controllers/webhook_controller_test.exs#L29

File: web/models/build_status_updater.ex (hunk 4 above)
Comment (jtslear) [inline web/models/build_status_updater.ex]: I guess this particular change isn't needed.  I'll revert this portion.

**Comment by drapergeek** [comment]
@jtslear Did you try this and verify it works on the test app I gave you?

Also I think you can change the router to use the `accepts json` instead of `html` and I think it will parse the JSON for you

**Comment by jtslear** [comment]
@drapergeek I don't think that'll work.  Travis is sending the hook in `application/x-www-form-urlencoded` format with a url encoded json sitting inside.

**Comment by jtslear** [comment]
And no, I didn't test it on heroku.

**Comment by jtslear** [comment]
I'm not sure how to test this inside of heroku.  But no errors, and a 200 was sent.

```
2016-03-02T04:07:40.879996+00:00 app[web.1]: 04:07:40.879 request_id=00e16ede-64ad-47c1-bbde-f5e6c5c3a6a4 [info] POST /webhook
2016-03-02T04:07:40.921144+00:00 app[web.1]: 04:07:40.920 request_id=00e16ede-64ad-47c1-bbde-f5e6c5c3a6a4 [info] Sent 200 in 40ms
```

**Comment by drapergeek** [comment]
Alright, tossed this up and ran it and this is the new output:

```
[debug] Processing by Bigben.WebhookController.create/2
  Parameters: %{"payload" => "{\"id\":113514168,\"repository\":{\"id\":1294851,\"name\":\"findit\",\"owner_name\":\"drapergeek\",\"url\":\"\"},\"number\":\"86\",\"config\":{\"sudo\":false,\"before_script\":[\"cp .sample.env .env\",\"bundle exec rake db:create\",\"bundle exec rake db:schema:load\"],\"cache\":\"bundler\",\"language\":\"ruby\",\"rvm\":[\"2.0.0\"],\"env\":[\"rspec spec/models\",\"rspec spec/features\"],\"notifications\":{\"webhooks\":\"https://ab4c63.ngrok.com/webhook\"},\".result\":\"configured\"},\"status\":0,\"result\":0,\"status_message\":\"Passed\",\"result_message\":\"Passed\",\"started_at\":\"2016-03-03T21:42:03Z\",\"finished_at\":\"2016-03-03T21:43:16Z\",\"duration\":140,\"build_url\":\"https://travis-ci.org/drapergeek/findit/builds/113514168\",\"commit_id\":32086634,\"commit\":\"c7a582a9f4017b06d04b96d0c7bf86a123835f48\",\"base_commit\":null,\"head_commit\":null,\"branch\":\"jd-try-to-add-multiple-jobs\",\"message\":\"Ngrok it up\",\"compare_url\":\"https://github.com/drapergeek/findit/compare/9d107c0a3dd4...c7a582a9f401\",\"committed_at\":\"2016-03-03T21:41:23Z\",\"author_name\":\"Jason Draper\",\"author_email\":\"jason@drapergeek.com\",\"committer_name\":\"Jason Draper\",\"committer_email\":\"jason@drapergeek.com\",\"matrix\":[{\"id\":113514169,\"repository_id\":1294851,\"parent_id\":113514168,\"number\":\"86.1\",\"state\":\"finished\",\"config\":{\"sudo\":false,\"before_script\":[\"cp .sample.env .env\",\"bundle exec rake db:create\",\"bundle exec rake db:schema:load\"],\"cache\":\"bundler\",\"language\":\"ruby\",\"rvm\":\"2.0.0\",\"env\":[\"rspec spec/models\"],\"notifications\":{\"webhooks\":\"https://ab4c63.ngrok.com/webhook\"},\".result\":\"configured\",\"group\":\"stable\",\"dist\":\"precise\",\"os\":\"linux\"},\"status\":0,\"result\":0,\"commit\":\"c7a582a9f4017b06d04b96d0c7bf86a123835f48\",\"branch\":\"jd-try-to-add-multiple-jobs\",\"message\":\"Ngrok it up\",\"compare_url\":\"https://github.com/drapergeek/findit/compare/9d107c0a3dd4...c7a582a9f401\",\"started_at\":\"2016-03-03T21:42:03Z\",\"finished_at\":\"2016-03-03T21:43:10Z\",\"committed_at\":\"2016-03-03T21:41:23Z\",\"author_name\":\"Jason Draper\",\"author_email\":\"jason@drapergeek.com\",\"committer_name\":\"Jason Draper\",\"committer_email\":\"jason@drapergeek.com\",\"allow_failure\":false},{\"id\":113514170,\"repository_id\":1294851,\"parent_id\":113514168,\"number\":\"86.2\",\"state\":\"finished\",\"config\":{\"sudo\":false,\"before_script\":[\"cp .sample.env .env\",\"bundle exec rake db:create\",\"bundle exec rake db:schema:load\"],\"cache\":\"bundler\",\"language\":\"ruby\",\"rvm\":\"2.0.0\",\"env\":[\"rspec spec/features\"],\"notifications\":{\"webhooks\":\"https://ab4c63.ngrok.com/webhook\"},\".result\":\"configured\",\"group\":\"stable\",\"dist\":\"precise\",\"os\":\"linux\"},\"status\":0,\"result\":0,\"commit\":\"c7a582a9f4017b06d04b96d0c7bf86a123835f48\",\"branch\":\"jd-try-to-add-multiple-jobs\",\"message\":\"Ngrok it up\",\"compare_url\":\"https://github.com/drapergeek/findit/compare/9d107c0a3dd4...c7a582a9f401\",\"started_at\":\"2016-03-03T21:42:03Z\",\"finished_at\":\"2016-03-03T21:43:16Z\",\"committed_at\":\"2016-03-03T21:41:23Z\",\"author_name\":\"Jason Draper\",\"author_email\":\"jason@drapergeek.com\",\"committer_name\":\"Jason Draper\",\"committer_email\":\"jason@drapergeek.com\",\"allow_failure\":false}],\"type\":\"push\",\"state\":\"passed\",\"pull_request\":false,\"pull_request_number\":null,\"pull_request_title\":null,\"tag\":null}"}
  Pipelines: []
[debug] SELECT b0."id", b0."name", b0."inserted_at", b0."updated_at" FROM "branches" AS b0 WHERE (b0."name" = $1) ["jd-try-to-add-multiple-jobs"] OK query=105.9ms queue=30.6ms
[debug] BEGIN [] OK query=0.4ms
[debug] INSERT INTO "branches" ("inserted_at", "updated_at", "name") VALUES ($1, $2, $3) RETURNING "id" [{{2016, 3, 3}, {21, 43, 17, 0}}, {{2016, 3, 3}, {21, 43, 17, 0}}, "jd-try-to-add-multiple-jobs"] OK query=1.2ms
[debug] COMMIT [] OK query=0.5ms
[error] #PID<0.409.0> running Bigben.Endpoint terminated
Server: ab4c63.ngrok.com:80 (http)
Request: POST /webhook
** (exit) an exception was raised:
    ** (ArgumentError) argument error
        :erlang.bit_size(113514168)
        (bigben) lib/travis/url_builder.ex:7: Bigben.Travis.UrlBuilder.build_url/1
        (bigben) lib/travis/total_time_calculator.ex:14: Bigben.Travis.TotalTimeCalculator.find_jobs/2
        (bigben) lib/travis/total_time_calculator.ex:8: Bigben.Travis.TotalTimeCalculator.calculate_total_time/1
        (bigben) web/models/build_status_updater.ex:18: Bigben.BuildStatusUpdater.add_total_time/1
        (bigben) web/models/build_status_updater.ex:11: Bigben.BuildStatusUpdater.update/1
        (bigben) web/controllers/webhook_controller.ex:9: Bigben.WebhookController.create/2
        (bigben) web/controllers/webhook_controller.ex:1: Bigben.WebhookController.action/2
        (bigben) web/controllers/webhook_controller.ex:1: Bigben.WebhookController.phoenix_controller_pipeline/2
        (bigben) lib/phoenix/router.ex:261: Bigben.Router.dispatch/2
        (bigben) web/router.ex:1: Bigben.Router.do_call/2
        (bigben) lib/plug/error_handler.ex:63: Bigben.Router.call/2
        (bigben) lib/bigben/endpoint.ex:1: Bigben.Endpoint.phoenix_pipeline/1
        (bigben) lib/plug/debugger.ex:93: Bigben.Endpoint."call (overridable 3)"/2
        (bigben) lib/phoenix/endpoint/render_errors.ex:34: Bigben.Endpoint.call/2
        (plug) lib/plug/adapters/cowboy/handler.ex:15: Plug.Adapters.Cowboy.Handler.upgrade/4
        (cowboy) src/cowboy_protocol.erl:442: :cowboy_protocol.execute/4


```

**Comment by drapergeek** [comment]
Added a new commit and now we're here:

```
[debug] Processing by Bigben.WebhookController.create/2
  Parameters: %{"payload" => "{\"id\":113515464,\"repository\":{\"id\":1294851,\"name\":\"findit\",\"owner_name\":\"drapergeek\",\"url\":\"\"},\"number\":\"87\",\"config\":{\"sudo\":false,\"before_script\":[\"cp .sample.env .env\",\"bundle exec rake db:create\",\"bundle exec rake db:schema:load\"],\"cache\":\"bundler\",\"language\":\"ruby\",\"rvm\":[\"2.0.0\"],\"env\":[\"rspec spec/models\",\"rspec spec/features\"],\"notifications\":{\"webhooks\":\"https://ab4c63.ngrok.com/webhook\"},\".result\":\"configured\"},\"status\":0,\"result\":0,\"status_message\":\"Passed\",\"result_message\":\"Passed\",\"started_at\":\"2016-03-03T21:47:04Z\",\"finished_at\":\"2016-03-03T21:48:11Z\",\"duration\":128,\"build_url\":\"https://travis-ci.org/drapergeek/findit/builds/113515464\",\"commit_id\":32086986,\"commit\":\"810bc6d87113e75229ee0c9a442a5bbf9c1df643\",\"base_commit\":null,\"head_commit\":null,\"branch\":\"jd-try-to-add-multiple-jobs\",\"message\":\"Another shot\",\"compare_url\":\"https://github.com/drapergeek/findit/compare/c7a582a9f401...810bc6d87113\",\"committed_at\":\"2016-03-03T21:45:57Z\",\"author_name\":\"Jason Draper\",\"author_email\":\"jason@drapergeek.com\",\"committer_name\":\"Jason Draper\",\"committer_email\":\"jason@drapergeek.com\",\"matrix\":[{\"id\":113515465,\"repository_id\":1294851,\"parent_id\":113515464,\"number\":\"87.1\",\"state\":\"finished\",\"config\":{\"sudo\":false,\"before_script\":[\"cp .sample.env .env\",\"bundle exec rake db:create\",\"bundle exec rake db:schema:load\"],\"cache\":\"bundler\",\"language\":\"ruby\",\"rvm\":\"2.0.0\",\"env\":[\"rspec spec/models\"],\"notifications\":{\"webhooks\":\"https://ab4c63.ngrok.com/webhook\"},\".result\":\"configured\",\"group\":\"stable\",\"dist\":\"precise\",\"os\":\"linux\"},\"status\":0,\"result\":0,\"commit\":\"810bc6d87113e75229ee0c9a442a5bbf9c1df643\",\"branch\":\"jd-try-to-add-multiple-jobs\",\"message\":\"Another shot\",\"compare_url\":\"https://github.com/drapergeek/findit/compare/c7a582a9f401...810bc6d87113\",\"started_at\":\"2016-03-03T21:47:04Z\",\"finished_at\":\"2016-03-03T21:48:05Z\",\"committed_at\":\"2016-03-03T21:45:57Z\",\"author_name\":\"Jason Draper\",\"author_email\":\"jason@drapergeek.com\",\"committer_name\":\"Jason Draper\",\"committer_email\":\"jason@drapergeek.com\",\"allow_failure\":false},{\"id\":113515466,\"repository_id\":1294851,\"parent_id\":113515464,\"number\":\"87.2\",\"state\":\"finished\",\"config\":{\"sudo\":false,\"before_script\":[\"cp .sample.env .env\",\"bundle exec rake db:create\",\"bundle exec rake db:schema:load\"],\"cache\":\"bundler\",\"language\":\"ruby\",\"rvm\":\"2.0.0\",\"env\":[\"rspec spec/features\"],\"notifications\":{\"webhooks\":\"https://ab4c63.ngrok.com/webhook\"},\".result\":\"configured\",\"group\":\"stable\",\"dist\":\"precise\",\"os\":\"linux\"},\"status\":0,\"result\":0,\"commit\":\"810bc6d87113e75229ee0c9a442a5bbf9c1df643\",\"branch\":\"jd-try-to-add-multiple-jobs\",\"message\":\"Another shot\",\"compare_url\":\"https://github.com/drapergeek/findit/compare/c7a582a9f401...810bc6d87113\",\"started_at\":\"2016-03-03T21:47:04Z\",\"finished_at\":\"2016-03-03T21:48:11Z\",\"committed_at\":\"2016-03-03T21:45:57Z\",\"author_name\":\"Jason Draper\",\"author_email\":\"jason@drapergeek.com\",\"committer_name\":\"Jason Draper\",\"committer_email\":\"jason@drapergeek.com\",\"allow_failure\":false}],\"type\":\"push\",\"state\":\"passed\",\"pull_request\":false,\"pull_request_number\":null,\"pull_request_title\":null,\"tag\":null}"}
  Pipelines: []
[debug] SELECT b0."id", b0."name", b0."inserted_at", b0."updated_at" FROM "branches" AS b0 WHERE (b0."name" = $1) ["jd-try-to-add-multiple-jobs"] OK query=112.3ms queue=18.6ms
[error] #PID<0.425.0> running Bigben.Endpoint terminated
Server: ab4c63.ngrok.com:80 (http)
Request: POST /webhook
** (exit) an exception was raised:
    ** (Protocol.UndefinedError) protocol Enumerable not implemented for nil
        (elixir) lib/enum.ex:1: Enumerable.impl_for!/1
        (elixir) lib/enum.ex:116: Enumerable.reduce/3
        (elixir) lib/enum.ex:1486: Enum.reduce/3
        (bigben) web/models/build_status_updater.ex:18: Bigben.BuildStatusUpdater.add_total_time/1
        (bigben) web/models/build_status_updater.ex:11: Bigben.BuildStatusUpdater.update/1
        (bigben) web/controllers/webhook_controller.ex:9: Bigben.WebhookController.create/2
        (bigben) web/controllers/webhook_controller.ex:1: Bigben.WebhookController.action/2
        (bigben) web/controllers/webhook_controller.ex:1: Bigben.WebhookController.phoenix_controller_pipeline/2
        (bigben) lib/phoenix/router.ex:261: Bigben.Router.dispatch/2
        (bigben) web/router.ex:1: Bigben.Router.do_call/2
        (bigben) lib/plug/error_handler.ex:63: Bigben.Router.call/2
        (bigben) lib/bigben/endpoint.ex:1: Bigben.Endpoint.phoenix_pipeline/1
        (bigben) lib/plug/debugger.ex:93: Bigben.Endpoint."call (overridable 3)"/2
        (bigben) lib/phoenix/endpoint/render_errors.ex:34: Bigben.Endpoint.call/2
        (plug) lib/plug/adapters/cowboy/handler.ex:15: Plug.Adapters.Cowboy.Handler.upgrade/4
```

**Comment by drapergeek** [comment]
Cleaned this up and merged here: https://github.com/drapergeek/bigben/commit/be2b98d6c22b3f15abfa0fd322583a596ae50283

**Pull request closed without merging**